tests/test_replay_buffer.py

```diff
@@ -18,52 +21,145 @@
+            correct_item.append(trans1)
+            rbuf.append(**trans1)
         self.assertEqual(len(rbuf), 1)
         s1 = rbuf.sample(1)
         self.assertEqual(len(s1), 1)
-        self.assertEqual(s1[0], trans1)
+        self.assertEqual(s1[0], correct_item)
 
         # Add two and sample two, which must be unique
+        correct_item2 = copy.deepcopy(correct_item)
         trans2 = dict(state=1, action=1, reward=2, next_state=3,
-                      next_action=4, is_state_terminal=True)
+                      next_action=4, is_state_terminal=False)
+        correct_item2.append(trans2)
         rbuf.append(**trans2)
         self.assertEqual(len(rbuf), 2)
         s2 = rbuf.sample(2)
         self.assertEqual(len(s2), 2)
-        if s2[0]['state'] == 0:
-            self.assertEqual(s2[0], trans1)
-            self.assertEqual(s2[1], trans2)
+        if s2[0][num_steps - 1]['state'] == 0:
+            self.assertEqual(s2[0], correct_item)
+            self.assertEqual(s2[1], correct_item2)
+        else:
+            self.assertEqual(s2[1], correct_item)
+            self.assertEqual(s2[0], correct_item2)
+
+    def test_append_and_terminate(self):
+        capacity = self.capacity
+        num_steps = self.num_steps
+        rbuf = replay_buffer.ReplayBuffer(capacity, num_steps)
+
+        self.assertEqual(len(rbuf), 0)
+
+        # Add one and sample one
+        for i in range(num_steps):
+            trans1 = dict(state=0, action=1, reward=2, next_state=3,
+                          next_action=4, is_state_terminal=False)
+            rbuf.append(**trans1)
+        self.assertEqual(len(rbuf), 1)
+        s1 = rbuf.sample(1)
+        self.assertEqual(len(s1), 1)
+
+        # Add two and sample two, which must be unique
+        trans2 = dict(state=1, action=1, reward=2, next_state=3,
+                      next_action=4, is_state_terminal=True)
+        rbuf.append(**trans2)
+        self.assertEqual(len(rbuf), self.num_steps + 1)
+        s2 = rbuf.sample(self.num_steps + 1)
+        self.assertEqual(len(s2), self.num_steps + 1)
+        if self.num_steps == 1:
+            if s2[0][0]['state'] == 0:
+                self.assertEqual(s2[1][0]['state'], 1)
+            else:
+                self.assertEqual(s2[1][0]['state'], 0)
         else:
-            self.assertEqual(s2[0], trans2)
-            self.assertEqual(s2[1], trans1)
+            for item in s2:
+                # e.g. if states are 0,0,0,1 then buffer looks like:
+                # [[0,0,0], [0, 0, 1], [0, 1], [1]]
+                if len(item) < self.num_steps:
+                    self.assertEqual(item[len(item) - 1]['state'], 1)
+                    for i in range(len(item) - 1):
+                        self.assertEqual(item[i]['state'], 0)
+                else:
+                    for i in range(len(item) - 1):
+                        self.assertEqual(item[i]['state'], 0)
+
+    def test_batch_experiences(self):
```

Comment:
Since `batch_experiences` is a function independent of `ReplayBuffer`, I think it is better to create a separate TestCase class for it.